src/audiovideocontroller/DefaultAudioVideoController.ts

```diff
@@ -733,28 +741,148 @@ export default class DefaultAudioVideoController
+
     const result = this.sessionStateController.perform(SessionStateControllerAction.Update, () => {
-      this.actionUpdate(true);
+      this.actionUpdateWithRenegotiation(true);
     });
     return (
       result === SessionStateControllerTransitionResult.Transitioned ||
       result === SessionStateControllerTransitionResult.DeferredTransition
     );
   }
 
+  // This function will try to use the diff between `this.meetingSessionContext.lastVideosToReceive`
+  // and `this.meetingSessionContext.videosToReceive` to determine if the changes can be accomplished
+  // through `SignalingClient.remoteVideoUpdate` rather then the full subscribe.
+  //
+  // It requires the caller to manage `this.meetingSessionContext.lastVideosToReceive`
+  // and `this.meetingSessionContext.videosToReceive` so that `this.meetingSessionContext.lastVideosToReceive`
+  // contains the stream IDs from either last time a subscribe was set, or last time this function was set.
+  //
+  // It will return true if succesful, if false the caller must fall back to a full renegotiation
+  private updateRemoteVideosFromLastVideosToReceive(): boolean {
+    const context = this.meetingSessionContext;
+    if (context.videosToReceive?.empty() || context.lastVideosToReceive?.empty()) {
+      return false;
+    }
+
+    // Check existence of all required dependencies and requisite functions
+    if (
+      !context.transceiverController ||
+      !context.transceiverController.getMidForStreamId ||
+      !context.transceiverController.setStreamIdForMid ||
+      !context.videosToReceive.forEach ||
+      !context.signalingClient.remoteVideoUpdate ||
+      !context.videoStreamIndex.overrideStreamIdMappings
+    ) {
+      return false;
+    }
+
+    let added: number[] = [];
+    const simulcastStreamUpdates: Map<number, number> = new Map();
+    let removed: number[] = [];
+
+    if (context.lastVideosToReceive === null) {
+      added = context.videosToReceive.array();
+    } else {
+      const index = context.videoStreamIndex;
+      context.videosToReceive.forEach((currentId: number) => {
+        if (context.lastVideosToReceive.contain(currentId)) {
+          return;
+        }
+
+        // Check if group ID exists in previous set (i.e. simulcast stream switch)
+        let foundUpdatedPreviousStreamId = false;
+        context.lastVideosToReceive.forEach((previousId: number) => {
+          if (index.StreamIdsInSameGroup(previousId, currentId)) {
+            simulcastStreamUpdates.set(previousId, currentId);
+            foundUpdatedPreviousStreamId = true;
+          }
+        });
+        if (!foundUpdatedPreviousStreamId) {
+          // Otherwise this must be a new stream
+          added.push(currentId);
+        }
+      });
+      removed = context.lastVideosToReceive.array().filter(idFromPrevious => {
+        const stillReceiving = context.videosToReceive
+          .array()
+          .some(idFromCurrent => idFromCurrent === idFromPrevious);
```

Comment:
I think you should implement `some` on `VideoStreamIdSet`, rather than allocating N+1 new arrays (line 822, line 824).